services/src/main/scala/cromwell/pipeline/service/GitLabProjectVersioning.scala

```diff
@@ -21,44 +20,57 @@ class GitLabProjectVersioning(httpClient: HttpClient, config: GitLabConfig)
     val versionValue = version.map(_.name).getOrElse(config.defaultFileVersion)
 
     httpClient
-      .put(
+      .put[Message, UpdateFileRequest](
         fileUrl,
-        payload = Json.stringify(
-          Json.toJson(UpdateFileRequest(versionValue, projectFile.content, versionValue))
-        ),
+        payload = UpdateFileRequest(versionValue, projectFile.content, versionValue),
         headers = config.token
       )
       .flatMap {
         case Response(HttpStatusCodes.OK, _, _) =>
           Future.successful(Right("Success update file"))
         case Response(HttpStatusCodes.BadRequest, _, _) =>
           httpClient
-            .post(
+            .post[Message, UpdateFileRequest](
               fileUrl,
-              payload = Json.stringify(
-                Json.toJson(UpdateFileRequest(config.defaultFileVersion, projectFile.content, "Init commit"))
-              ),
+              payload = UpdateFileRequest(config.defaultFileVersion, projectFile.content, "Init commit"),
               headers = config.token
             )
             .map {
-              case Response(HttpStatusCodes.OK, _, _) => Right("Create new file")
-              case Response(_, body, _)               => Left(VersioningException(body))
+              case Response(HttpStatusCodes.OK, _, _)         => Right("Create new file")
+              case Response(_, NegativeResponseBody(body), _) => Left(VersioningException(body))
+              case _ =>
+                Left(
+                  VersioningException(
+                    "status is not Ok, but response is not negative, check status and possible positive responses"
+                  )
+                )
             }
-        case Response(_, body, _) => Future.successful(Left(VersioningException(body)))
+        case Response(_, body, _) =>
+          body match {
+            case PositiveResponseBody(bodyInfo) =>
+              Future.successful(
+                Left(VersioningException("status codes processing may not correctly implemented: " + bodyInfo))
+              )
+            case NegativeResponseBody(error) => Future.successful(Left(VersioningException(error)))
+          }
+
       }
   }
 
   override def updateFiles(project: Project, projectFiles: ProjectFiles)(
     implicit ec: ExecutionContext
   ): AsyncResult[List[String]] = ???
 
-  override def createRepository(project: Project)(implicit ec: ExecutionContext): AsyncResult[Project] =
+  override def createRepository(project: Project)(implicit ec: ExecutionContext): AsyncResult[Project] = {
+
+    import cromwell.pipeline.datastorage.dto.Project._
```

Comment:
Move it to other imports at the top of the file.

Reply:
done